docs/source/config/config.rst
Nginx and WSGI
--------------

Production |st2| installations use `nginx <http://nginx.org/en/>`_ for SSL termination,
serving static content of WebUI,
and running st2auth and st2api as WSGI apps via gunicorn/uwsgi. StackStorm nginx configurations
can be found at ``/etc/nginx/sites-enabled/st2*.conf``.


``st2auth`` and ``st2api`` can also run in by a built-in simple Python server. This is used for development and strongly discouraged for any production. Be aware that some settings in /etc/st2.conf are only effective when running in development mode, and don't apply when running under WSGI servers. Refer to the comments in
:github_st2:`st2.conf.sample <conf/st2.conf.sample>`.

Configure MongoDB
-----------------

StackStorm requires a connection to MongoDB to operate.


In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [database]
    host = <MongoDB host>
    port = <MongoDB server port>
    db_name = <User define database name, usually st2>
    username = <username for db login>
    password = <password for db login>

The ``username`` and ``password`` properties are optional.

Configure RabbitMQ
------------------

StackStorm uses RabbitMQ for messaging between its services.

In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [messaging]
    url = <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_HOST:#RMQ_PORT/#RMQ_VHOST>

The ``#RMQ_VHOST`` property is optional and can be left blank.

StackStorm also supports `RabbitMQ cluster <https://www.rabbitmq.com/clustering.html>`_.

In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [messaging]
    cluster_urls = <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_1:#RMQ_PORT/#RMQ_VHOST>,
                   <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_2:#RMQ_PORT/#RMQ_VHOST>,
                   <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_3:#RMQ_PORT/#RMQ_VHOST>


* To understand more about setting up a RabbitMQ cluster - https://www.rabbitmq.com/clustering.html
* RabbitMQ HA guide - https://www.rabbitmq.com/ha.html


.. _config-configure-ssh:

Configure SSH
-------------

To run actions on remote hosts, |st2| uses SSH. It is advised to configure identity file based SSH access on all remote hosts.

|st2| ssh user and a path to SSH key are set in ``/etc/st2/st2.conf``. During installation, ``st2_deploy.sh`` script configures ssh on the local box for a user `stanley`.

Follow these steps on a remote box to setup `stanley` user on remote boxes.

.. code-block:: bash

    useradd stanley
    mkdir -p /home/stanley/.ssh
    chmod 0700 /home/stanley/.ssh

    # generate ssh keys on StackStorm box and copy over public key into remote box.
    ssh-keygen -f /home/stanley/.ssh/stanley_rsa -P ""
    cp ${KEY_LOCATION}/stanley_rsa.pub /home/stanley/.ssh/stanley_rsa.pub

    # authorize key-base acces.
    cat /home/stanley/.ssh/stanley_rsa.pub >> /home/stanley/.ssh/authorized_keys
    chmod 0600 /home/stanley/.ssh/authorized_keys
    chown -R stanley:stanley /home/stanley
    echo "stanley    ALL=(ALL)       NOPASSWD: SETENV: ALL" >> /etc/sudoers.d/st2

    # ensure requiretty is not set to default in the /etc/sudoers file.

To verify do the following from the |st2| box

.. code-block:: bash

    # ssh should not require a password since the key is already provided
    ssh -i /home/stanley/.ssh/stanley_rsa stanely@host.example.com

    # make sure that no password is prompted.
    sudo su

SSH Troubleshooting
~~~~~~~~~~~~~~~~~~~

* Validate that passwordless SSH configuration works fine for the destination. Assuming default user `stanley`:

    .. code-block:: bash

        sudo ssh -i /home/stanley/.ssh/stanley_rsa -t stanley@host.example.com uname -a

Using SSH config
~~~~~~~~~~~~~~~~

StackStorm allows loading of the SSH config file local to the system user. This is a configurable option and to
enable add following to ``/etc/st2/st2.conf``

.. code-block:: bash

    [ssh_runner]
    use_ssh_config = True
    ...

SUDO Access
-----------

StackStorm's ``shell`` actions -  ``local-shell-cmd``, ``local-shell-script``, ``remote-shell-cmd``, ``remote-shell-script``- are performed by a special user. By default, this user is named ``stanley`` and that is configurable via :github_st2:`st2.conf <conf/st2.prod.conf>`.

.. note:: `stanley` user requires the following access:

    * Sudo access to all boxes on which script action will run.
    * SETENV option needs to be set for all the commands. This way environment variables which are
      available to the local runner actions will also be available when user executes local runner
      action under a different user or with root privileges.
    * As some actions require sudo privileges password-less sudo access to all boxes.

One way of setting up passwordless sudo is perform the below operation on each remote box.

.. code-block:: bash

    echo "stanley    ALL=(ALL)       NOPASSWD: SETENV: ALL" >> /etc/sudoers.d/st2


Configure Logging
-----------------

By default, the logs can be found in ``/var/log/st2``.

* With the standard logging setup you will notice files like ``st2*.log`` and
  ``st2*.audit.log`` in the log folder.

* Per component logging configuration can be found in ``/etc/st2*/logging.conf``.
  Those files use `Python logging configuration format <https://docs.python.org/2/library/logging.config.html#configuration-file-format>`_.
  If you desire to change location of the log files, the paths and other
  settings can be modified in these files.

* By default, log rotation is handled via logrotate. Default log rotation config
  (:github_st2:`logrotate.conf <conf/logrotate.conf>`) is included with all the
  package based installations. If you want Python services instead of logrotate
  to handle the log rotation for you, you can update the logging configs as
  shown below:

  .. code-block:: ini

      [handler_fileHandler]
      class=handlers.RotatingFileHandler
      level=DEBUG
      formatter=verboseConsoleFormatter
      args=("logs/st2api.log", "a", 100000000, 5)

  In this case the log file will be rotated when it reaches 100000000 bytes (100
  MB) and a maximum of 5 old log files will be kept. For more information, see
  `RotatingFileHandler <https://docs.python.org/2/library/logging.handlers.html#rotatingfilehandler>`_
  docs.

* To configure logging with syslog, grab the configuration and follow
  instructions at :github_contrib:`st2contrib/extra/syslog <extra/syslog>`

* Check out LogStash configuration and Kibana dashboard for pretty logging and
  audit at :github_contrib:`st2contrib/extra/logstash <extra/logstash>`

  logrotate log rotation

Configure Mistral
-----------------
There are a number of configurable options available under the mistral section in ``/etc/st2/st2.conf``. If the mistral section is not provided, default values will be used. By default, all Keystone related options are unset and |st2| will not pass any credential for authentication to Mistral. Please refer to OpenStack and Mistral documentation for Keystone setup.

+-----------------------+--------------------------------------------------------+
| options               | description                                            |
+=======================+========================================================+
| v2_base_url           | Mistral API v2 root endpoint                           |
+-----------------------+--------------------------------------------------------+
| retry_exp_msec        | Multiplier for the exponential backoff.                |
+-----------------------+--------------------------------------------------------+
| retry_exp_max_msec    | Max time for each set of backoff.                      |
+-----------------------+--------------------------------------------------------+
| retry_stop_max_msec   | Max time to stop retrying.                             |
+-----------------------+--------------------------------------------------------+
| keystone_username     | Username for authentication with OpenStack Keystone.   |
+-----------------------+--------------------------------------------------------+
| keystone_password     | Password for authentication with OpenStack Keystone.   |
+-----------------------+--------------------------------------------------------+
| keystone_project_name | OpenStack project scope.                               |
+-----------------------+--------------------------------------------------------+
| keystone_auth_url     | v3 Auth URL for OpenStack Keystone.                    |
+-----------------------+--------------------------------------------------------+

::

    # Example with basic options. The v2_base_url is set to http://workflow.example.com:8989/v2.
    # On connection error, the following configuration sets up the action runner to retry 
    # connecting to mistral for up to 10 minutes. The retries is setup to be exponential for
    # 5 minutes. So in this case, there will be two sets of exponential retries during
    # the 10 minutes.

    [mistral]
    v2_base_url = http://workflow.example.com:8989/v2
    retry_exp_msec = 1000
    retry_exp_max_msec = 300000
    retry_stop_max_msec = 600000

::

    # Example with auth options.

    [mistral]
    v2_base_url = http://workflow.example.com:8989/v2
    retry_exp_msec = 1000
    retry_exp_max_msec = 300000
    retry_stop_max_msec = 600000
    keystone_username = mistral
    keystone_password = pass123
    keystone_project_name = default
    keystone_auht_url = http://identity.example.com:5000/v3


Authentication
--------------

Please refer to :doc:`../authentication` to learn details of authentication, integrations with
various identity providers, managing API tokens.

Configure ChatOps
-----------------

|st2| brings native two-way ChatOps support. To learn more about ChatOps, and how to configure it manually, please refer to :ref:`Configuration section under ChatOps <chatops-configuration>`.





docs/source/install/config.rst
Configuration
==============

.. note:: If you are using the "all in one" :doc:`/install/index`, all configurations are already setup.

|st2| configuration file is at :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>`

Configure MongoDB
-----------------

StackStorm requires a connection to MongoDB to operate.


In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [database]
    host = <MongoDB host>
    port = <MongoDB server port>
    db_name = <User define database name, usually st2>
    username = <username for db login>
    password = <password for db login>

.. note:: The username and password properties are optional.

Configure RabbitMQ
------------------

StackStorm uses RabbitMQ for messaging between its services.

In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [messaging]
    url = <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_HOST:#RMQ_PORT/#RMQ_VHOST>

.. note:: The #RMQ_VHOST property is optional and can be left blank.

StackStorm also supports RabbitMQ cluster.

In :github_st2:`/etc/st2/st2.conf <conf/st2.prod.conf>` include the following section :

.. code-block:: bash

    [messaging]
    cluster_urls = <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_1:#RMQ_PORT/#RMQ_VHOST>,
                   <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_2:#RMQ_PORT/#RMQ_VHOST>,
                   <amqp://#RMQ_USER:#RMQ_PASSWD@#RMQ_NODE_3:#RMQ_PORT/#RMQ_VHOST>


* To understand more about setting up a RabbitMQ cluster - https://www.rabbitmq.com/clustering.html
* RabbitMQ HA guide - https://www.rabbitmq.com/ha.html

SUDO Access
-----------

By default, all actions run by |st2| are performed by a single user. Typically, this user is named
``stanley`` and that is configurable via :github_st2:`st2.conf <conf/st2.prod.conf>`.

.. note:: `stanley` user requires the following access -

    * Sudo access to all boxes on which action will run.
    * SETENV option needs to be set for all the commands. This way environment variables which are
      available to the local runner actions will also be available when user executes local runner
      action under a different user or with root privileges.
    * As some actions require sudo privileges password-less sudo access to all boxes.

One option of setting up passwordless sudo is perform the below operation on each remote box.

.. code-block:: bash

    echo "stanley    ALL=(ALL)       NOPASSWD: SETENV: ALL" >> /etc/sudoers.d/st2

.. _config-configure-ssh:

Configure SSH
-------------

To run actions on remote hosts, |st2| uses `Fabric <http://www.fabfile.org/>`_. It is required to configure identity file based SSH access on all remote hosts.

|st2| ssh user and a path to SSH key are set in ``/etc/st2/st2.conf``. During installation, ``st2_deploy.sh`` script configures ssh on the local box for a user `stanley`.

Follow these steps on a remote box to setup `stanley` user on remote boxes.

.. code-block:: bash

    useradd stanley
    mkdir -p /home/stanley/.ssh
    chmod 0700 /home/stanley/.ssh

    # generate ssh keys on StackStorm box and copy over public key into remote box.
    ssh-keygen -f /home/stanley/.ssh/stanley_rsa -P ""
    cp ${KEY_LOCATION}/stanley_rsa.pub /home/stanley/.ssh/stanley_rsa.pub

    # authorize key-base acces.
    cat /home/stanley/.ssh/stanley_rsa.pub >> /home/stanley/.ssh/authorized_keys
    chmod 0600 /home/stanley/.ssh/authorized_keys
    chown -R stanley:stanley /home/stanley
    echo "stanley    ALL=(ALL)       NOPASSWD: SETENV: ALL" >> /etc/sudoers.d/st2

    # ensure requiretty is not set to default in the /etc/sudoers file.

To verify do the following from the |st2| box

.. code-block:: bash

    # ssh should not require a password since the key is already provided
    ssh -i /home/stanley/.ssh/stanley_rsa stanely@host.example.com

    # make sure that no password is prompted.
    sudo su

SSH Troubleshooting
~~~~~~~~~~~~~~~~~~~

* Validate that passwordless SSH configuration works fine for the destination. Assuming default user `stanley`:

    .. code-block:: bash

        sudo ssh -i /home/stanley/.ssh/stanley_rsa -t stanley@host.example.com uname -a

Using SSH config
~~~~~~~~~~~~~~~~

StackStorm allows loading of the SSH config file local to the system user. This is a configurable option and to
enable add following to ``/etc/st2/st2.conf``

.. code-block:: bash

    [ssh_runner]
    use_ssh_config = True
    ...


Configure Logging
-----------------

By default, the logs can be found in ``/var/log/st2``.

* With the standard logging setup you will notice files like ``st2*.log`` and
  ``st2*.audit.log`` in the log folder.

* Per component logging configuration can be found in ``/etc/st2*/logging.conf``.
  Those files use `Python logging configuration format <https://docs.python.org/2/library/logging.config.html#configuration-file-format>`_.
  If you desire to change location of the log files, the paths and other
  settings can be modified in these files.

* By default, log rotation is handled via logrotate. Default log rotation config
  (:github_st2:`logrotate.conf <conf/logrotate.conf>`) is included with all the
  package based installations. If you want Python services instead of logrotate
  to handle the log rotation for you, you can update the logging configs as
  shown below:

  .. code-block:: ini

      [handler_fileHandler]
      class=handlers.RotatingFileHandler
      level=DEBUG
      formatter=verboseConsoleFormatter
      args=("logs/st2api.log", , "a", 100000000, 5)

  In this case the log file will be rotated when it reaches 100000000 bytes (100
  MB) and a maximum of 5 old log files will be kept. For more information, see
  `RotatingFileHandler <https://docs.python.org/2/library/logging.handlers.html#rotatingfilehandler>`_
  docs.

* To configure logging with syslog, grab the configuration and follow
  instructions at :github_contrib:`st2contrib/extra/syslog <extra/syslog>`

* Check out LogStash configuration and Kibana dashboard for pretty logging and
  audit at :github_contrib:`st2contrib/extra/logstash <extra/logstash>`

  logrotate log rotation

Configure Mistral
-----------------
There are a number of configurable options available under the mistral section in ``/etc/st2/st2.conf``. If the mistral section is not provided, default values will be used. By default, all Keystone related options are unset and |st2| will not pass any credential for authentication to Mistral. Please refer to OpenStack and Mistral documentation for Keystone setup.

+-----------------------+--------------------------------------------------------+
| options               | description                                            |
+=======================+========================================================+
| v2_base_url           | Mistral API v2 root endpoint                           |
+-----------------------+--------------------------------------------------------+
| max_attempts          | Max attempts to reconnect on connection error.         |
+-----------------------+--------------------------------------------------------+
| retry_wait            | Number of seconds to wait inbetween reconnection.      |
+-----------------------+--------------------------------------------------------+
| keystone_username     | Username for authentication with OpenStack Keystone.   |
+-----------------------+--------------------------------------------------------+
| keystone_password     | Password for authentication with OpenStack Keystone.   |
+-----------------------+--------------------------------------------------------+
| keystone_project_name | OpenStack project scope.                               |
+-----------------------+--------------------------------------------------------+
| keystone_auth_url     | v3 Auth URL for OpenStack Keystone.                    |
+-----------------------+--------------------------------------------------------+

::

    # Example with basic options.

    [mistral]
    v2_base_url = http://workflow.example.com:8989/v2
    max_attempts = 180
    retry_wait = 5

::

    # Example with auth options.

    [mistral]
    v2_base_url = http://workflow.example.com:8989/v2
    max_attempts = 180
    retry_wait = 5
    keystone_username = mistral
    keystone_password = pass123
    keystone_project_name = default
    keystone_auht_url = http://identity.example.com:5000/v3

Authentication
--------------
Please refer to the main :doc:`/../authentication` section.

Sample configuration file
-------------------------

A sample config file with all the configuration options can be found at :github_st2:`st2.conf.sample <conf/st2.conf.sample>`.

Serve WebUI files from the API server
-------------------------------------

By default, static WebUI files are served by python SimpleHTTPServer. This means you can
access the web interface by going to ``http://<api host>:8080/``.

For production deployments, we encourage you use nginx, Apache or a similar
dedicated web server to serve those static files.

.. include:: /engage.rst


CHANGELOG.rst
Changelog
=========

in development
--------------

* Allow user to pass ``env`` parameter to ``packs.setup_virtualenv`` and ``packs.install``
  action.

  This comes handy if user wants pip to use an HTTP(s) proxy (HTTP_PROXY and HTTPS_PROXY
  environment variable) when installing pack dependencies. (new feature)
* Ability to view causation chains in Trace. This helps reduce the noise when using Trace to
  identify specific issues. (new-feature)
* Filter Trace components by model types to only view ActionExecutions, Rules or TriggerInstances.
  (new-feature)
* Include ref of the most meaningful object in each trace component. (new-feature)
* Ability to hide trigger-instance that do not yield a rule enforcement. (new-feature)
* Change the rule list columns in the CLI from ref, pack, description and enabled to ref,
  trigger.ref, action.ref and enabled. This aligns closer the UI and also brings important
  information front and center. (improvement)
* Action and Trigger filters for rule list (new-feature)
* Add missing logrotate config entry for ``st2auth`` service. #2294 [Vignesh Terafast]
* Support for object already present in the DB for ``st2-rule-tester`` (improvement)
* Add ``--register-fail-on-failure`` flag to ``st2-register-content`` script. If this flag is
  provided, the script will fail and exit with non-zero status code if registering some resource
  fails. (new feature)
* Add a missing ``get_logger`` method to the `MockSensorService``. This method now returns an
  instance of ``Mock`` class which allows user to assert that a particular message has been
  logged. [Tim Ireland, Tomaz Muraus]
* Introduce a new ``abandoned`` state that is applied to executions that we cannot guarantee as
  completed. Typically happen when an actionrunner currently running some executions quits or is
  killed via TERM.
* Add new ``st2garbagecollector`` service which periodically deletes old data from the database
  as configured in the config. By default, no old data is deleted unless explicitly configured in
  the config.
* All published variables can be available in the result of ActionChain execution under the
  ``published`` property if ``display_published`` property is specified.
* Allow user to specify TTL when creating datastore item using CLI with the ``--ttl`` option.
  (improvement)
* Fix validation error when None is passed explicitly to an optional argument on action
  execution. (bug fix)
* Deprecated ``params`` action attribute in the action chain definition in favor of the new
  ``parameters`` attribute. (improvement)
* Fix action parameters validation so that only a selected set of attributes can be overriden for
  any runner parameters. (bug fix)
* Fix type in the headers parameter for the http-request runner. (bug fix)
* Fix runaway action triggers caused by state miscalculation for mistral workflow. (bug fix) 

1.2.0 - December 07, 2015
-------------------------

* Refactor retries in the Mistral action runner to use exponential backoff. Configuration options
  for Mistral have changed. (improvement)
* Add SSH bastion host support to the paramiko SSH runner. Utilizes same connection parameters as
  the targeted box. (new feature, improvement) #2144, #2150 [Logan Attwood]
* Update action chain runner so it performs on-success and on-error task name validation during
  pre_run time. This way common errors such as typos in the task names can be spotted early on
  since there is no need to wait for the run time.
* Change ``headers`` and ``params`` ``core.http`` action paramer type from ``string`` to
  ``object``.
* Don't allow action parameter ``type`` attribute to be an array since rest of the code doesn't
  support parameters with multiple types. (improvement)
* Fix trigger parameters validation for system triggers during rule creation - make sure we
  validate the parameters before creating a TriggerDB object. (bug fix)
* Update local runner so all the commands which are executed as a different user and result in
  using sudo set $HOME variable to the home directory of the target user. (improvement)
* Fix a bug with a user inside the context of the live action which was created using alias
  execution endpoint incorrectly being set to the system user (``stanley``) instead of the
  authenticated user which triggered the execution. (bug fix)
* Include state_info for Mistral workflow and tasks in the action execution result. (improvement)
* Introduce a new ``timeout`` action execution status which represents an action execution
  timeout. Previously, executions which timed out had status set to ``failure``. Keep in mind
  that timeout is just a special type of a failure. (new feature)
* ``--debug`` flag no longer implies profiling mode. If you want to enable profiling mode, you need
  to explicitly pass ``--profile`` flag to the binary. To reproduce the old behavior, simply pass
  both flags to the binary - ``--debug --profile``.
* Fix policy loading and registering - make sure we validate policy parameters against the
  parameters schema when loading / registering policies. (bug fix, improvement)
* Fix policy trigger for action execution cancellation. (bug fix)
* Improve error reporting for static error in ActionChain definition e.g. incorrect reference
  in default etc. (improvement)
* Fix action chain so it doesn't end up in an infinite loop if an action which is part of the chain
  is canceled. (bug fix)
* Allow jinja templating to be used in ``message`` and ``data`` field for notifications.(new feature)
* Add tools for purging executions (also, liveactions with it) and trigger instances older than
  certain UTC timestamp from the db in bulk.
* Fix json representation of trace in cli. (bug fix)
* Introducing `noop` runner and `core.noop` action. Returns consistent success in a WF regardless of
  user input. (new feature)
* Add missing indexes on trigger_instance_d_b collection. (bug fix)

st2common/st2common/config.py
# Licensed to the StackStorm, Inc ('StackStorm') under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os

from oslo_config import cfg

from st2common.constants.system import VERSION_STRING


def do_register_opts(opts, group=None, ignore_errors=False):
    try:
        cfg.CONF.register_opts(opts, group=group)
    except:
        if not ignore_errors:
            raise


def do_register_cli_opts(opt, ignore_errors=False):
    # TODO: This function has broken name, it should work with lists :/
    if not isinstance(opt, (list, tuple)):
        opts = [opt]
    else:
        opts = opt

    try:
        cfg.CONF.register_cli_opts(opts)
    except:
        if not ignore_errors:
            raise


def register_opts(ignore_errors=False):
    auth_opts = [
        cfg.BoolOpt('enable', default=True, help='Enable authentication middleware.'),
        cfg.IntOpt('token_ttl', default=86400, help='Access token ttl in seconds.')
    ]
    do_register_opts(auth_opts, 'auth', ignore_errors)

    rbac_opts = [
        cfg.BoolOpt('enable', default=False, help='Enable RBAC.'),
    ]
    do_register_opts(rbac_opts, 'rbac', ignore_errors)

    system_user_opts = [
        cfg.StrOpt('user',
                   default='stanley',
                   help='Default system user.'),
        cfg.StrOpt('ssh_key_file',
                   default='/home/vagrant/.ssh/stanley_rsa',
                   help='SSH private key for the system user.')
    ]
    do_register_opts(system_user_opts, 'system_user', ignore_errors)

    schema_opts = [
        cfg.IntOpt('version', default=4, help='Version of JSON schema to use.'),
        cfg.StrOpt('draft', default='http://json-schema.org/draft-04/schema#',
                   help='URL to the JSON schema draft.')
    ]
    do_register_opts(schema_opts, 'schema', ignore_errors)

    system_opts = [
        cfg.StrOpt('base_path', default='/opt/stackstorm',
                   help='Base path to all st2 artifacts.'),
        cfg.ListOpt('admin_users', default=[],
                    help='A list of usernames for users which should have admin privileges')
    ]
    do_register_opts(system_opts, 'system', ignore_errors)

    system_packs_base_path = os.path.join(cfg.CONF.system.base_path, 'packs')
    content_opts = [
        cfg.StrOpt('system_packs_base_path', default=system_packs_base_path,
                   help='Path to the directory which contains system packs.'),
        cfg.StrOpt('packs_base_paths', default=None,
                   help='Paths which will be searched for integration packs.')
    ]
    do_register_opts(content_opts, 'content', ignore_errors)

    db_opts = [
        cfg.StrOpt('host', default='0.0.0.0', help='host of db server'),
        cfg.IntOpt('port', default=27017, help='port of db server'),
        cfg.StrOpt('db_name', default='st2', help='name of database'),
        cfg.StrOpt('username', help='username for db login'),
        cfg.StrOpt('password', help='password for db login'),
        cfg.IntOpt('connection_retry_max_delay_m', help='Connection retry total time (minutes).',
                   default=3),
        cfg.IntOpt('connection_retry_backoff_max_s', help='Connection retry backoff max (seconds).',
                   default=10),
        cfg.IntOpt('connection_retry_backoff_mul', help='Backoff multiplier (seconds).',
                   default=1)
    ]
    do_register_opts(db_opts, 'database', ignore_errors)

    messaging_opts = [
        # It would be nice to be able to deprecate url and completely switch to using
        # url. However, this will be a breaking change and will have impact so allowing both.
        cfg.StrOpt('url', default='amqp://guest:guest@127.0.0.1:5672//',
                   help='URL of the messaging server.'),
        cfg.ListOpt('cluster_urls', default=[],
                    help='URL of all the nodes in a messaging service cluster.')
    ]
    do_register_opts(messaging_opts, 'messaging', ignore_errors)

    syslog_opts = [
        cfg.StrOpt('host', default='127.0.0.1',
                   help='Host for the syslog server.'),
        cfg.IntOpt('port', default=514,
                   help='Port for the syslog server.'),
        cfg.StrOpt('facility', default='local7',
                   help='Syslog facility level.'),
        cfg.StrOpt('protocol', default='udp',
                   help='Transport protocol to use (udp / tcp).')
    ]
    do_register_opts(syslog_opts, 'syslog', ignore_errors)

    log_opts = [
        cfg.ListOpt('excludes', default='',
                    help='Exclusion list of loggers to omit.'),
        cfg.BoolOpt('redirect_stderr', default=False,
                    help='Controls if stderr should be redirected to the logs.'),
        cfg.BoolOpt('mask_secrets', default=True,
                    help='True to mask secrets in the log files.')
    ]
    do_register_opts(log_opts, 'log', ignore_errors)

    # Common API options
    api_opts = [
        cfg.StrOpt('host', default='0.0.0.0', help='StackStorm API server host'),
        cfg.IntOpt('port', default=9101, help='StackStorm API server port')
    ]
    do_register_opts(api_opts, 'api', ignore_errors)

    # Common auth options
    auth_opts = [
        cfg.StrOpt('api_url', default=None,
                   help='Base URL to the API endpoint excluding the version')
    ]
    do_register_opts(auth_opts, 'auth', ignore_errors)

    # Common options (used by action runner and sensor container)
    action_sensor_opts = [
        cfg.BoolOpt('enable', default=True,
                    help='Whether to enable or disable the ability to post a trigger on action.'),
    ]
    do_register_opts(action_sensor_opts, group='action_sensor')

    # Coordination options
    coord_opts = [
        cfg.StrOpt('url', default=None, help='Endpoint for the coordination server.'),
        cfg.IntOpt('lock_timeout', default=60, help='TTL for the lock if backend suports it.')
    ]
    do_register_opts(coord_opts, 'coordination', ignore_errors)

    # Mistral options
    mistral_opts = [
        cfg.StrOpt('v2_base_url', default='http://127.0.0.1:8989/v2', help='v2 API root endpoint.'),
        cfg.IntOpt('retry_exp_msec', default=1000, help='Multiplier for the exponential backoff.'),
        cfg.IntOpt('retry_exp_max_msec', default=300000, help='Max time for each set of backoff.'),
        cfg.IntOpt('retry_stop_max_msec', default=600000, help='Max time to stop retrying.'),
        cfg.StrOpt('keystone_username', default=None, help='Username for authentication.'),
        cfg.StrOpt('keystone_password', default=None, help='Password for authentication.'),
        cfg.StrOpt('keystone_project_name', default=None, help='OpenStack project scope.'),
        cfg.StrOpt('keystone_auth_url', default=None, help='Auth endpoint for Keystone.'),

        cfg.StrOpt('api_url', default=None, help=('URL Mistral uses to talk back to the API.'
            'If not provided it defaults to public API URL. Note: This needs to be a base '
            'URL without API version (e.g. http://127.0.0.1:9101)'))
    ]
    do_register_opts(mistral_opts, group='mistral', ignore_errors=ignore_errors)

    # Common CLI options
    debug = cfg.BoolOpt('debug', default=False,
        help='Enable debug mode. By default this will set all log levels to DEBUG.')
    profile = cfg.BoolOpt('profile', default=False,
        help=('Enable profile mode. In the profile mode all the MongoDB queries and related '
              'profile data are logged.'))
    use_debugger = cfg.BoolOpt('use-debugger', default=True,
        help='Enables debugger. Note that using this option changes how the '
             'eventlet library is used to support async IO. This could result in '
             'failures that do not occur under normal operation.')

    cli_opts = [debug, profile, use_debugger]
    do_register_cli_opts(cli_opts, ignore_errors=ignore_errors)


def parse_args(args=None):
    register_opts()
    cfg.CONF(args=args, version=VERSION_STRING)
